<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.8"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>scran_norm: Scaling normalization of single-cell count data</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">scran_norm
   </div>
   <div id="projectbrief">Scaling normalization of single-cell data</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.8 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
</div><!-- top -->
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div><div class="header">
  <div class="headertitle"><div class="title">Scaling normalization of single-cell count data </div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p><a class="anchor" id="md__2github_2workspace_2README"></a> <img src="https://github.com/libscran/scran_norm/actions/workflows/run-tests.yaml/badge.svg" alt="Unit tests" style="pointer-events: none;" class="inline"/> <img src="https://github.com/libscran/scran_norm/actions/workflows/doxygenate.yaml/badge.svg" alt="Documentation" style="pointer-events: none;" class="inline"/> <a href="https://codecov.io/gh/libscran/scran_norm"><img src="https://codecov.io/gh/libscran/scran_norm/graph/badge.svg?token=JWV0I4WJX2" alt="Codecov" style="pointer-events: none;" class="inline"/></a></p>
<h1>Overview</h1>
<p>This repository contains functions to perform scaling normalization and log-transformation of a gene-by-cell count matrix. Normalization removes per-cell scaling biases such as differences in capture efficiency and sequencing depth, while the log-transformation provides some variance stabilization and allows differences in values to be interpreted as log-fold changes. The code itself was originally derived from the <a href="https://bioconductor.org/packages/scran"><b>scran</b> R package</a>, factored out into a separate C++ library for easier re-use.</p>
<h1>Quick start</h1>
<p>Given a measure of the per-cell scaling bias - for example, the sum of counts for a cell - we can convert them into centered size factors:</p>
<div class="fragment"><div class="line"><span class="preprocessor">#include &quot;<a class="code" href="scran__norm_8hpp.html">scran_norm/scran_norm.hpp</a>&quot;</span></div>
<div class="line"> </div>
<div class="line"><span class="comment">// Assuming that counts is a std::shared_ptr&lt;tatami::Matrix&gt;</span></div>
<div class="line">std::vector&lt;double&gt; bias = tatami_stats::sums::by_column(counts.get()); </div>
<div class="line"> </div>
<div class="line">scran_norm::CenterSizeFacotrsOptions copt;</div>
<div class="line">scran_norm::center_size_factors(bias.size(), bias.data(), NULL, copt);</div>
<div class="line"> </div>
<div class="line"><span class="comment">// &#39;bias&#39; is now centered at unity and can be used as size factors.</span></div>
<div class="line"><span class="keyword">auto</span>&amp; size_factors = bias;</div>
<div class="ttc" id="ascran__norm_8hpp_html"><div class="ttname"><a href="scran__norm_8hpp.html">scran_norm.hpp</a></div><div class="ttdoc">Scaling normalization of single-cell data.</div></div>
</div><!-- fragment --><p>Alternatively, in the presence of blocks, we adjust our centering so that the mean size factor in each block is no less than 1. This avoids inflated variances from applying small size factors to low-coverage blocks.</p>
<div class="fragment"><div class="line">scran_norm::center_size_factors_blocked(</div>
<div class="line">    bias.size(), </div>
<div class="line">    bias.data(), </div>
<div class="line">    block.data(), </div>
<div class="line">    NULL, </div>
<div class="line">    copt</div>
<div class="line">);</div>
</div><!-- fragment --><p>If our size factors might contain invalid values (i.e., zero, negative, or non-finite), we can sanitize them prior to the construction of the log-normalized matrix:</p>
<div class="fragment"><div class="line"><a class="code hl_struct" href="structscran__norm_1_1SanitizeSizeFactorsOptions.html">scran_norm::SanitizeSizeFactorsOptions</a> sopt;</div>
<div class="line">sopt.<a class="code hl_variable" href="structscran__norm_1_1SanitizeSizeFactorsOptions.html#a341805c7650ee93db13050621d1505d6">handle_zero</a> = scran_norm::SanitizeAction::SANITIZE;</div>
<div class="line">sopt.<a class="code hl_variable" href="structscran__norm_1_1SanitizeSizeFactorsOptions.html#acaf0c77542462c508c14c30994e8d07d">handle_infinite</a> = scran_norm::SanitizeAction::SANITIZE;</div>
<div class="line">scran_norm::sanitize_size_factors(size_factors.size(), size_factors.data(), sopt);</div>
<div class="line"><span class="comment">// Gets rid of any zero and infinite values in &#39;size_factors&#39;.</span></div>
<div class="ttc" id="astructscran__norm_1_1SanitizeSizeFactorsOptions_html"><div class="ttname"><a href="structscran__norm_1_1SanitizeSizeFactorsOptions.html">scran_norm::SanitizeSizeFactorsOptions</a></div><div class="ttdoc">Options for sanitize_size_factors().</div><div class="ttdef"><b>Definition</b> sanitize_size_factors.hpp:142</div></div>
<div class="ttc" id="astructscran__norm_1_1SanitizeSizeFactorsOptions_html_a341805c7650ee93db13050621d1505d6"><div class="ttname"><a href="structscran__norm_1_1SanitizeSizeFactorsOptions.html#a341805c7650ee93db13050621d1505d6">scran_norm::SanitizeSizeFactorsOptions::handle_zero</a></div><div class="ttdeci">SanitizeAction handle_zero</div><div class="ttdef"><b>Definition</b> sanitize_size_factors.hpp:154</div></div>
<div class="ttc" id="astructscran__norm_1_1SanitizeSizeFactorsOptions_html_acaf0c77542462c508c14c30994e8d07d"><div class="ttname"><a href="structscran__norm_1_1SanitizeSizeFactorsOptions.html#acaf0c77542462c508c14c30994e8d07d">scran_norm::SanitizeSizeFactorsOptions::handle_infinite</a></div><div class="ttdeci">SanitizeAction handle_infinite</div><div class="ttdef"><b>Definition</b> sanitize_size_factors.hpp:174</div></div>
</div><!-- fragment --><p>Finally, we convert our <a href="https://github.com/tatami-inc/tatami"><code>tatami::Matrix</code></a> of counts into a log-transformed normalized matrix:</p>
<div class="fragment"><div class="line"><a class="code hl_struct" href="structscran__norm_1_1NormalizeCountsOptions.html">scran_norm::NormalizeCountsOptions</a> lopts;</div>
<div class="line"><span class="keyword">auto</span> logcounts = scran_norm::normalize_counts(counts, size_factors, lopt);</div>
<div class="ttc" id="astructscran__norm_1_1NormalizeCountsOptions_html"><div class="ttname"><a href="structscran__norm_1_1NormalizeCountsOptions.html">scran_norm::NormalizeCountsOptions</a></div><div class="ttdoc">Options for normalize_counts().</div><div class="ttdef"><b>Definition</b> normalize_counts.hpp:20</div></div>
</div><!-- fragment --><p>Check out the <a href="https://libscran.github.io/scran_norm">reference documentation</a> for more details.</p>
<h1>Building projects</h1>
<h2>CMake with <code>FetchContent</code></h2>
<p>If you're using CMake, you just need to add something like this to your <code>CMakeLists.txt</code>:</p>
<div class="fragment"><div class="line">include(FetchContent)</div>
<div class="line"> </div>
<div class="line">FetchContent_Declare(</div>
<div class="line">  scran_norm</div>
<div class="line">  GIT_REPOSITORY https://github.com/libscran/scran_norm</div>
<div class="line">  GIT_TAG master # or any version of interest</div>
<div class="line">)</div>
<div class="line"> </div>
<div class="line">FetchContent_MakeAvailable(scran_norm)</div>
</div><!-- fragment --><p>Then you can link to <b>scran_norm</b> to make the headers available during compilation:</p>
<div class="fragment"><div class="line"># For executables:</div>
<div class="line">target_link_libraries(myexe libscran::scran_norm)</div>
<div class="line"> </div>
<div class="line"># For libaries</div>
<div class="line">target_link_libraries(mylib INTERFACE libscran::scran_norm)</div>
</div><!-- fragment --><h2>CMake with <code>find_package()</code></h2>
<div class="fragment"><div class="line">find_package(libscran_scran_norm CONFIG REQUIRED)</div>
<div class="line">target_link_libraries(mylib INTERFACE libscran::scran_norm)</div>
</div><!-- fragment --><p>To install the library, use:</p>
<div class="fragment"><div class="line">mkdir build &amp;&amp; cd build</div>
<div class="line">cmake .. -DSCRAN_NORM_TESTS=OFF</div>
<div class="line">cmake --build . --target install</div>
</div><!-- fragment --><p>By default, this will use <code>FetchContent</code> to fetch all external dependencies. If you want to install them manually, use <code>-DSCRAN_NORM_FETCH_EXTERN=OFF</code>. See the tags in <a href="extern/CMakeLists.txt"><code>extern/CMakeLists.txt</code></a> to find compatible versions of each dependency.</p>
<h2>Manual</h2>
<p>If you're not using CMake, the simple approach is to just copy the files in <code>include/</code> - either directly or with Git submodules - and include their path during compilation with, e.g., GCC's <code>-I</code>. This requires the external dependencies listed in <a href="extern/CMakeLists.txt"><code>extern/CMakeLists.txt</code></a>, which also need to be made available during compilation. </p>
</div></div><!-- PageDoc -->
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.8
</small></address>
</body>
</html>
